webui/react/src/pages/F_ExpList/F_ExperimentList.tsx

```diff
@@ -161,95 +172,114 @@ const F_ExperimentList: React.FC<Props> = ({ project }) => {
+    // observable only works on initial loads -- the settings store will be
+    // fully loaded when navigating to the page in-router, so we check if we
+    // need to subscribe at all.
+    const initFormset = (ps: Loadable<ProjectSettings | null>) => {
+      ps.forEach((s) => {
+        if (!s?.filterset) return;
+        const formSetValidation = IOFilterFormSet.decode(JSON.parse(s.filterset));
+        if (isLeft(formSetValidation)) {
+          handleError(formSetValidation.left, {
+            publicSubject: 'Unable to initialize filterset from settings',
+          });
+        } else {
+          formStore.init(formSetValidation.right);
+        }
+        cleanup = formStore.asJsonString.subscribe(() => {
+          resetPagination();
+          const loadableFormset = formStore.formset.get();
+          Loadable.forEach(loadableFormset, (formSet) =>
+            updateSettings({ filterset: JSON.stringify(formSet) }),
+          );
+        });
       });
+    };
+    if (projectSettingsObs.get().isLoaded) {
+      initFormset(projectSettingsObs.get());
     } else {
-      const formset = formSetValidation.right;
-      formStore.init(formset);
+      cleanup = projectSettingsObs.subscribe((ps) => {
+        if (ps.isLoaded) {
+          cleanup();
+          initFormset(ps);
+        }
+      });
     }
-  }, [settings.filterset, isLoadingSettings]);
+    return () => cleanup?.();
+  }, [projectSettingsObs, resetPagination, updateSettings]);
 
   const [isLoading, setIsLoading] = useState(true);
   const [error] = useState(false);
   const [canceler] = useState(new AbortController());
 
-  const colorMap = useGlasbey(settings.selectedExperiments);
-  const { height: containerHeight, width: containerWidth } = useResize(contentRef);
-  const height =
-    containerHeight - 2 * parseInt(getThemeVar('strokeWidth')) - (isPagedView ? 40 : 0);
-  const [scrollPositionSetCount] = useState(observable(0));
-
-  const selectedExperimentIds: Set<number> = useMemo(() => {
-    return isLoadingSettings ? new Set() : new Set(settings.selectedExperiments);
-  }, [isLoadingSettings, settings.selectedExperiments]);
-
-  const excludedExperimentIds: Set<number> = useMemo(() => {
-    return isLoadingSettings ? new Set() : new Set(settings.excludedExperiments);
-  }, [isLoadingSettings, settings.excludedExperiments]);
-
-  useEffect(() => {
-    if (isLoading) return;
-
-    const selectedIds = new Set(selectedExperimentIds);
+  // partition experiment list into not selected/selected with indices and experiments so we only iterate the result list once
+  const [excludedExperimentIds, selectedExperimentIds]: [
+    Map<number, { index: number; experiment: ExperimentItem }>,
+    Map<number, { index: number; experiment: ExperimentItem }>,
+  ] = useMemo(() => {
+    if (isLoadingSettings) {
+      return [new Map(), new Map()];
+    }
+    const selectedMap = new Map();
+    const excludedMap = new Map();
```

Comment:
super nit: you could define the selectedMap and excludedMap before the conditional and just return them if empty. saves you from constructing two additional maps

Reply:
Also it would be good to move the map types to here so that the rest of the code in the useMemo can benefit from the typechecking instead of having `Map<any, any>`

Reply:
good point!